File: frontend/src/lib/platform.ts
import type { Source } from "../api";

export type Platform = "youtube" | "lesswrong" | "web" | "pdf" | "arxiv";

const LESSWRONG_HOSTS = new Set(["lesswrong.com", "alignmentforum.org"]);

export function platformFromSource(source: Source): Platform {
  if (source.platform) {
    return source.platform as Platform;
  }
  if (source.source_type === "youtube") return "youtube";
  if (source.source_type === "pdf") return "pdf";
  if (source.source_type === "arxiv") return "arxiv";
  if (source.url) {
    try {
      const host = new URL(source.url).hostname.replace(/^www\./, "").toLowerCase();
      if (LESSWRONG_HOSTS.has(host)) return "lesswrong";
    } catch {
      /* ignore invalid URLs */
    }
  }
  return "web";
}

export const MEDIUM_OPTIONS: { id: Platform; label: string }[] = [
  { id: "youtube", label: "YouTube" },
  { id: "lesswrong", label: "LessWrong" },
  { id: "web", label: "Web" },
  { id: "pdf", label: "PDF" },
  { id: "arxiv", label: "ArXiv" },
];

File: src/meridian/ingest/platform.py
from __future__ import annotations

from urllib.parse import urlparse

LESSWRONG_HOSTS = frozenset({"lesswrong.com", "alignmentforum.org"})


def platform_for_source(*, url: str | None, source_type: str) -> str:
    if source_type == "youtube":
        return "youtube"
    if source_type == "pdf":
        return "pdf"
    if source_type == "arxiv":
        return "arxiv"
    if url:
        host = urlparse(url).netloc.removeprefix("www.").lower()
        if host in LESSWRONG_HOSTS:
            return "lesswrong"
    if source_type == "web":
        return "web"
    return source_type
